### ISSUE TITLE
[Bug]: "Blocked by dependencies" and "Blocked mono-repo" when updating `hydra`

### ISSUE THREAD

**Issue by edgar-vincent** [progfolio/elpaca#194]
### Elpaca Version

Elpaca 7a34cd3 HEAD -> master, origin/master, origin/HEAD
installer:      0.5
emacs-version:  GNU Emacs 29.1 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8)
git --version:  git version 2.42.0

### Operating System

Arch Linux

### Description

1. Install `hydra`.
 ``` emacs-lisp
 (elpaca-test
  :init (elpaca hydra)
  :interactive t)
```
2. In `elpaca-test` instance of Emacs:
`M-x elpaca-update-all`

3. Content of `elpaca-log`:
```
hydra                          blocked              Blocked by dependencies: (lv)                                                    23.355217
lv                             blocked              Blocked mono-repo. waiting for hydra                                             23.355240
```

Menu item for hydra:
```
( :package   "hydra"
  :repo      "abo-abo/hydra"
  :fetcher   github
  :files    
  (:defaults (:exclude "lv.el")))
```

Menu item for lv:
```
( :package   "lv"
  :repo      "abo-abo/hydra"
  :fetcher   github
  :files    
  (:defaults "lv.el"))
```

There seems to be a race between the two recipes which prevents the package from being updated.

Thanks a lot for your excellent work!

**Comment by progfolio** [comment]
>  ``` emacs-lisp
>  (elpaca-test
>   :init (elpaca hydra)
>   :interactive t)
> ```

Thank you for taking the time to fill out a bug report and providing a perfect reproduction case.

> There seems to be a race between the two recipes which prevents the package from being updated.

There's an bug in `elpaca-update-all` when package's are being queued.
The "mono-repo" which hosts both hydra and lv should only be fetched once.
`elpaca-update-all` correctly detects this, and blocks lv when queuing the updates. 
However, it doesn't account for the case where there is a dependency between the sibling libraries in the mono-repo.

Mono-repos in general have been a thorn in my side since I started working on Elpaca.
There's a WIP branch which should fix this specific case.
See the following test case:

<details open><summary>Test Case</summary>

[How to run this test?](https://github.com/progfolio/elpaca/wiki/Troubleshooting#the-elpaca-test-macro)

```emacs-lisp
(elpaca-test
  :early-init (setq elpaca-menu-functions '(elpaca-menu-melpa))
  :ref "refactor/messaging"
  :init (elpaca hydra)
  (elpaca-wait)
  (elpaca-update-all)
  (elpaca-wait)
  (elpaca-test-log "#latest hydra\\\\|lv"))
```

</details>
<details><summary>Host Env</summary>

<table>
<tr><td>elpaca</td><td>11b561d HEAD -> master, origin/master, origin/HEAD</td>
<tr><td>isntaller</td><td>0.5</td>
<tr><td>emacs</td><td>GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.18.0) of 2023-10-09</td>
<tr><td>git</td><td>git version 2.42.0</td>
</table>
</details>

<details><summary>Output</summary>

```emacs-lisp
  INFO     Scraping files for loaddefs... 
  INFO     Scraping files for loaddefs...done
  GEN      ../elpaca-autoloads.el
Cloning into '/tmp/elpaca.GleY3d/elpaca/repos/elpaca'...
Switched to a new branch 'refactor/messaging'
branch 'refactor/messaging' set up to track 'origin/refactor/messaging'.
Checking /tmp/elpaca.GleY3d/elpaca/repos/elpaca...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/elpaca-info.el...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/elpaca-log.el...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/elpaca-manager.el...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/elpaca-menu-elpa.el...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/elpaca-menu-melpa.el...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/elpaca-menu-org.el...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/elpaca-process.el...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/elpaca-test.el...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/elpaca-ui.el...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/elpaca.el...
Checking /tmp/elpaca.GleY3d/elpaca/repos/elpaca/doc...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/doc/early-init.el...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/doc/init.el...
Checking /tmp/elpaca.GleY3d/elpaca/repos/elpaca/extensions...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/extensions/elpaca-use-package.el...
Checking /tmp/elpaca.GleY3d/elpaca/repos/elpaca/images...
Checking /tmp/elpaca.GleY3d/elpaca/repos/elpaca/test...
Compiling /tmp/elpaca.GleY3d/elpaca/repos/elpaca/test/elpaca-test.el...
Done (Total of 11 files compiled, 3 skipped in 4 directories)

Downloading MELPA recipes...
Downloading MELPA recipes...100%
elpaca-log: "#latest hydra\\\\|lv"
"
hydra                          fetching-remotes     POST git-upload-pack (416 bytes)                                                 03.548181
hydra                          fetching-remotes     From https://github.com/abo-abo/hydra                                            03.614624
hydra                          fetching-remotes      = [up to date]      master     -> origin/master                                 03.617478
hydra                          fetching-remotes     Remotes fetched                                                                  03.622503
hydra                          merging              Merging updates                                                                  03.635060
hydra                          merging              Already up to date.                                                              03.637976
lv                             unblocked            Unblocked by: hydra                                                              03.642233
lv                             linking              Linking build files                                                              03.644143
lv                             linking              Build files linked                                                               03.645905
lv                             autoloads            Generating autoloads: /tmp/elpaca.GleY3d/elpaca/builds/lv                        03.647494
hydra                          finished             ✓ 0.246 secs                                                                     03.649406
lv                             autoloads              INFO     Scraping files for loaddefs...                                        03.762959
lv                             autoloads              INFO     Scraping files for loaddefs...done                                    03.764997
lv                             autoloads              GEN      lv-autoloads.el                                                       03.765683
lv                             autoloads            Autoloads Generated                                                              03.768384
lv                             byte-compilation     Byte compiling                                                                   03.769178
lv                             byte-compilation     Checking /tmp/elpaca.GleY3d/elpaca/builds/lv...                                  03.833618
lv                             byte-compilation     Compiling /tmp/elpaca.GleY3d/elpaca/builds/lv/lv-autoloads.el...                 03.837973
lv                             byte-compilation     Compiling /tmp/elpaca.GleY3d/elpaca/builds/lv/lv.el...                           03.838739
lv                             byte-compilation     Done (Total of 1 file compiled, 1 skipped)                                       03.845697
lv                             byte-compilation     Byte compilation complete                                                        03.846408
lv                             info                 Compiling Info files                                                             03.847162
lv                             info                 Info source files not found                                                      03.847780
lv                             finished             ✓ 1.058 secs                                                                     03.848449
"

 Test Env

Elpaca 20f690d HEAD -> refactor/messaging, origin/refactor/messaging
installer:      0.5
emacs-version:  GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.18.0)
 of 2023-10-09
git --version:  git version 2.42.0
```
</details>

It's not quite correct either, though.
If hydra is already up to date, lv shouldn't need to be rebuilt.
There are other corner cases that need to be handled as well.

I plan on addressing all of this by refactoring the way packages communicate with each other. 
I'll let you know when that's finished.
In the mean time, you should be able to unblock the mutually blocked packages by updating one of them. e.g. `M-x elpaca-update hydra`.

**Comment by edgar-vincent** [comment]
Thank you so much for your detailed comment -- and for all your work on this excellent package.

**Comment by progfolio** [comment]
> Thank you so much for your detailed comment -- and for all your work on this
> excellent package.

Thank you for the kind words.

I've merged some changes which should take care of this issue.
See the test case below:


<details open><summary>Test Case</summary>

[How to run this test?](https://github.com/progfolio/elpaca/wiki/Troubleshooting#the-elpaca-test-macro)

```emacs-lisp
(elpaca-test
  :early-init (setq elpaca-menu-functions '(elpaca-menu-melpa))
  :init (elpaca hydra)
  (elpaca-wait)
  (elpaca-update-all 'fetch)
  (elpaca-wait)
  (elpaca-test-log "#latest #unique"))
```

</details>
<details><summary>Host Env</summary>

<table>
<tr><td>elpaca</td><td>518e7b6 HEAD -> master, origin/master, origin/HEAD</td>
<tr><td>isntaller</td><td>0.6</td>
<tr><td>emacs</td><td>GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.18.0) of 2023-11-02</td>
<tr><td>git</td><td>git version 2.42.0</td>
</table>
</details>

<details><summary>Output</summary>

```emacs-lisp
  INFO     Scraping files for loaddefs... 
  INFO     Scraping files for loaddefs...done
  GEN      ../elpaca-autoloads.el
Cloning into '/tmp/elpaca.5WeLTw/elpaca/repos/elpaca'...
Your branch is up to date with 'origin/master'.
Checking /tmp/elpaca.5WeLTw/elpaca/repos/elpaca...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/elpaca-info.el...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/elpaca-log.el...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/elpaca-manager.el...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/elpaca-menu-elpa.el...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/elpaca-menu-melpa.el...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/elpaca-menu-org.el...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/elpaca-process.el...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/elpaca-test.el...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/elpaca-ui.el...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/elpaca.el...
Checking /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/doc...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/doc/early-init.el...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/doc/init.el...
Checking /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/extensions...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/extensions/elpaca-use-package.el...
Checking /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/images...
Checking /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/test...
Compiling /tmp/elpaca.5WeLTw/elpaca/repos/elpaca/test/elpaca-test.el...
Done (Total of 11 files compiled, 3 skipped in 4 directories)

Downloading MELPA recipes...
Downloading MELPA recipes...100%
elpaca-log: "#latest #unique"
"
lv                             finished             ✓ 0.261 secs                                                                     03.407957
hydra                          finished             ✓ 0.259 secs                                                                     03.408503
elpaca                         finished             ✓ 0.262 secs                                                                     03.410436
"

 Test Env

Elpaca 518e7b6 HEAD -> master, origin/master, origin/HEAD
installer:      0.6
emacs-version:  GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.18.0)
 of 2023-11-02
git --version:  git version 2.42.0
```
</details>

Testing is appreciated. Note #195 when using `elpaca-update-all` interactively. Thanks again.

**Comment by liebkne** [comment]
Not sure if it's related, when I do a fresh installation, the progress is blocked and the window is frozen here forever

<img width="744" alt="image" src="https://github.com/progfolio/elpaca/assets/5475262/2b6dbeff-8c2d-4629-b191-0b62b173c315">

**Comment by progfolio** [comment]
> Not sure if it's related, when I do a fresh installation, the progress is blocked and the window is frozen here forever
> <img alt="image" width="744" src="https://user-images.githubusercontent.com/5475262/280199702-2b6dbeff-8c2d-4629-b191-0b62b173c315.png">

please include the output of `M-x elpaca-version`

**Comment by liebkne** [comment]
Here it is
```
Elpaca 2c1a667 HEAD -> master, origin/master, origin/HEAD
installer:      0.6
emacs-version:  GNU Emacs 29.1 (build 1, x86_64-apple-darwin23.0.0, Carbon Version 170 AppKit 2487)
 of 2023-10-28
git --version:  git version 2.42.1
```

**Comment by progfolio** [comment]
> Here it is
> 
> ```
> Elpaca 2c1a667 HEAD -> master, origin/master, origin/HEAD
> installer:      0.6
> emacs-version:  GNU Emacs 29.1 (build 1, x86_64-apple-darwin23.0.0, Carbon Version 170 AppKit 2487)
>  of 2023-10-28
> git --version:  git version 2.42.1
> ```
> 
> I did a `elpaca-update-all` and found the packages somehow broken so I deleted the `elpaca` directory and did a fresh installation and encountered this.

This looks like a related, but separate issue.
Please open a new bug report and be sure to include the relevant portion of your configuration (or a link to your init files).
I will take a look at it in the morning. Thank you.

**Comment by edgar-vincent** [comment]
Thanks a lot, @progfolio. I will test this when I have access to my PC on Tuesday.

3 Nov 2023 06:26:52 Nicholas Vollmer ***@***.***>:

**Comment by progfolio** [comment]
@edgar-vincent Thank you. Sounds good.

@liebkne 
I've merged the changes to the master branch.
Everything is working for me with the test cases I came up with, but testing is always appreciated.

You both should be able to `M-x elpaca-update elpaca` and restart to test. Thanks again.

**Comment by progfolio** [comment]
> Thanks a ton @progfolio, it all seems to be working fine now.

Glad to hear it! Thank you for testing.

**Comment by progfolio** [comment]
Considering this resolved. @liebkne if the issue you were running into persists after updating, please let me know and we can re-open. Thanks.

**Comment by liebkne** [comment]
@progfolio The issue seems gone now, thanks!
